c7n/resources/aws.py

```diff
@@ -324,6 +341,90 @@ def handle_context_missing(self):
         so default to disabling context missing output.
         """
 
+    # Annotate any segments/subsegments with their thread ids.
+    def put_segment(self, segment):
+        if getattr(segment, 'thread_id', None) is None:
+            segment.thread_id = threading.get_ident()
+        super().put_segment(segment)
+
+    def put_subsegment(self, subsegment):
+        if getattr(subsegment, 'thread_id', None) is None:
+            subsegment.thread_id = threading.get_ident()
+        super().put_subsegment(subsegment)
+
+    # Override since we're not just popping the end of the stack, we're removing
+    # the thread subsegment fromxo the array by identity.
+    def end_subsegment(self, end_time):
+        subsegment = self.get_trace_entity()
+        if self._is_subsegment(subsegment):
+            subsegment.close(end_time)
+            self._local.entities.remove(subsegment)
+            return True
+        else:
+            log.warning("No subsegment to end.")
+            return False
+
+    # Override get trace identity, any worker thread will find its own subsegment
+    # on the stack, else will use the main thread's sub/segment
+    def get_trace_entity(self):
+        tid = threading.get_ident()
+        entities = self._local.get('entities', ())
+        for s in reversed(entities):
+            if s.thread_id == tid:
+                return s
+        for s in reversed(entities):
+            if s.thread_id == self._main_tid:
```

Comment:
if we never expect the thread pool to advance sans a main thread segment closing we can just add this into the above loop